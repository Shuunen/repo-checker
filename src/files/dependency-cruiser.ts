import { blue, green } from 'shuutils'
import { FileBase } from '../file.ts'

// eslint-disable-next-line no-restricted-syntax, jsdoc/require-jsdoc
export class DependencyCruiserFile extends FileBase {
  /**
   * Start the dependency cruiser file check
   */
  public async start() {
    if (this.data.isUsingKnip) return
    const isUsingDepCruiser = this.test(this.data.isUsingDependencyCruiser, `use ${green('https://knip.dev')} or ${blue('https://www.npmjs.com/package/dependency-cruiser')}`, true)
    /* c8 ignore next 4 */
    if (!isUsingDepCruiser) return
    const hasJsFile = await this.fileExists('.dependency-cruiser.js')
    const hasCjsFile = await this.fileExists('.dependency-cruiser.cjs')
    this.test(hasJsFile || hasCjsFile, 'has a .dependency-cruiser config file')
  }
}
