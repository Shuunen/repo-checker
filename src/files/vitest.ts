import { FileBase } from '../file.ts'

// eslint-disable-next-line no-export
export class VitestFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  // eslint-disable-next-line max-lines-per-function
  public async start() {
    if (!this.data.isUsingVitest) return
    const hasMtsFile = await this.fileExists('vitest.config.mts')
    const hasJsFile = await this.fileExists('vitest.config.js')
    const hasTsFile = await this.fileExists('vitest.config.ts')
    const hasMjsFile = await this.fileExists('vitest.config.mjs')
    if (hasMtsFile) await this.inspectFile('vitest.config.mts')
    /* c8 ignore next 3 */
    if (hasJsFile) await this.inspectFile('vitest.config.js')
    if (hasTsFile) await this.inspectFile('vitest.config.ts')
    if (hasMjsFile) await this.inspectFile('vitest.config.mjs')
    this.couldContains('a pool thread setting', /pool: 'threads'/u)
  }
}
