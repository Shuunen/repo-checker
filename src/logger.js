/* eslint-disable no-console */
const fs = require('fs')
const chalk = require('chalk')

const pkg = require('../package.json')

class Logger {
  get indent () {
    return '    '.repeat(this.indentLevel)
  }

  get date () {
    return new Date().toISOString().replace('T', ' ').split('.')[0]
  }

  constructor (filePath) {
    this.file = fs.createWriteStream(filePath, { flags: 'a' })
    this.indentLevel = 0
  }

  async _write (...args) {
    this.file.write(args.join(' ') + '\n')
  }

  async setIndentLevel (level) {
    this.indentLevel = level
  }

  debug (...args) {
    return this._write(this.indent, 'üîπ', ...args)
  }

  info (...args) {
    console.log(this.indent, '‚¨ú ', ...args)
    return this._write(this.indent, '‚¨ú', ...args)
  }

  error (...args) {
    console.error(chalk.redBright(this.indent, '‚ùå ', ...args))
    return this._write(this.indent, '‚ùå', ...args)
  }

  warn (...args) {
    console.log(chalk.yellowBright(this.indent, '‚ö†Ô∏è ', ...args))
    return this._write(this.indent, '‚ö†Ô∏è', ...args)
  }

  success (...args) {
    console.log(chalk.green(this.indent, '‚úîÔ∏è ', ...args))
    return this._write(this.indent, '‚úîÔ∏è', ...args)
  }

  test (ok, msg, justWarn) {
    if (ok) {
      this.success(msg)
    } else if (justWarn) {
      this.warn(msg)
    } else {
      this.error(msg)
    }
  }

  fix (...args) {
    console.log(chalk.blueBright(this.indent, '‚¨ú ', ...args))
    return this._write(this.indent, '‚¨ú', ...args)
  }

  line () {
    console.log('')
    return this._write('')
  }

  start () {
    this.line()
    this._write(`‚¨áÔ∏è--- Entry from ${this.date} ---‚¨áÔ∏è`)
    this.info(`${pkg.name} v${pkg.version} is starting`)
    return this.line()
  }

  end () {
    this.line()
    this.info(pkg.name, 'has finished')
    return this.line()
  }
}

const log = new Logger(pkg.config.logFile)

module.exports = log
