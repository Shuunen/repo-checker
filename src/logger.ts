import { bgBlue, black, Logger } from 'shuutils'

class ExtendedLogger extends Logger {
  public fix (...stuff: readonly string[]): void {
    const items = Array.from(stuff)
    items.push(bgBlue(black('[ fixed ]')))
    this.info(...items)
  }

  public unknownError (error: unknown): void {
    if (error instanceof Error) { this.error(error.message); return }
    if (typeof error === 'string') { this.error(error); return }
    this.error('Unknown error format, could not log it')
  }
}

export const log = new ExtendedLogger()
