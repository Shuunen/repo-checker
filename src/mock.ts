/* eslint-disable jsdoc/require-jsdoc */
import path from 'node:path'
import { clone, sleep } from 'shuutils'
import type { Indicators } from './check.ts'
import type { FileBase } from './file.ts'
import { join } from './utils.ts'

const fileBaseKeysToDelete: (keyof FileBase)[] = ['fileExists', 'checkFileExists', 'folderPath', 'inspectFile', 'updateFile', 'initFile']
const indicatorsKeys: ReadonlyArray<keyof Indicators> = ['failed', 'passed', 'warnings'] as const

function cleanStringForSnap(input: string) {
  return (
    input
      // replace repo-check-1-40 by repo-check-x-yy
      .replace(/repo-check-\d-\d+/gu, 'repo-check-x-yy')
      // replace package-json-main-file-size-63ko by package-json-main-file-size-XYZko
      .replace(/package-json-main-file-size-\d+ko/gu, 'package-json-main-file-size-XYZko')
  )
}

// eslint-disable-next-line no-restricted-syntax
export function cleanUnknownValueForSnap<Type>(input: Type): Type {
  /* c8 ignore next 8 */
  /* eslint-disable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
  if (typeof input === 'string') return cleanStringForSnap(input) as Type
  if (Array.isArray(input)) return input.map(item => cleanUnknownValueForSnap(item)) as Type
  if (typeof input === 'object') return JSON.parse(cleanStringForSnap(JSON.stringify(input))) as Type
  // eslint-disable-next-line no-restricted-syntax
  throw new Error(`cleanInstanceValueForSnap: unknown type ${typeof input}`)
  /* eslint-enable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
}

export async function promiseValue<Type>(value: Type) {
  await sleep(1)
  return value
}

export async function promiseTrue() {
  await sleep(1)
  return true
}

export async function promiseFalse() {
  await sleep(1)
  return false
}

export async function promiseVoid() {
  await sleep(1)
}

// eslint-disable-next-line unicorn/prefer-module
export const sourceFolder = path.dirname(__filename)

export const vueProjectFolder = join(sourceFolder, 'mocks', 'vueProject')

export const tsProjectFolder = join(sourceFolder, 'mocks', 'tsProject')

export const mocksProjectsFolder = join(sourceFolder, 'mocks')

export function cleanIndicatorsForSnap(record?: Readonly<Record<string, unknown>>) {
  if (!record) return record
  const clean = clone(record)
  for (const key of indicatorsKeys) if (key in clean) clean[key] = cleanUnknownValueForSnap(clean[key])
  return clean
}

// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
export function cleanInstanceForSnap(instance: Readonly<FileBase>, ...bonusKeys: readonly string[]) {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const clean = clone<Partial<FileBase>>(instance) as Record<string, unknown>
  const keys = [...fileBaseKeysToDelete, ...bonusKeys] // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
  for (const key of keys) if (key in clean) delete clean[key]
  return cleanIndicatorsForSnap(clean)
}
