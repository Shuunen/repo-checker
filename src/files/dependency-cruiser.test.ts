import { expect, it } from 'vitest'
import { log } from '../logger'
import { cleanInstanceForSnap, promiseFalse } from '../mock'
import { DependencyCruiserFile } from './dependency-cruiser'

it('dependency cruiser config missing file', async () => {
  log.disable()
  const instance = new DependencyCruiserFile()
  instance.checkFileExists = promiseFalse
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})
