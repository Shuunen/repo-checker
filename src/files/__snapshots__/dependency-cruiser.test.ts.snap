// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dependency cruiser config missing file 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isCli": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": false,
    "isUsingBiome": false,
    "isUsingBun": false,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingKnip": false,
    "isUsingNyc": false,
    "isUsingOxc": false,
    "isUsingPreact": false,
    "isUsingReact": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingV8": false,
    "isUsingVitest": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [
    "use-32mhttps-knip-dev39m-or-34mhttps-www-npmjs-com-package-dependency-cruiser39m",
  ],
}
`;
